plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TaskRunner } from '@backstage/backend-tasks';
+import { AwsS3Integration } from '@backstage/integration';
+import {
+  EntityProvider,
+  EntityProviderConnection,
+  LocationSpec,
+  locationSpecToLocationEntity,
+} from '@backstage/plugin-catalog-backend';
+import { AwsCredentials } from '../credentials/AwsCredentials';
+import { AwsS3Config } from './types';
+import { S3 } from 'aws-sdk';
+import { ListObjectsV2Output } from 'aws-sdk/clients/s3';
+import * as uuid from 'uuid';
+import { Logger } from 'winston';
+
+// TODO: event-based updates using S3 events (+ queue like SQS)?
+/**
+ * Provider which discovers catalog files (any name) within an S3 bucket.
+ *
+ * Use `AwsS3EntityProviderFactory.fromConfig(...)` to create instances.
+ *
+ * @public
+ */
+export class AwsS3EntityProvider implements EntityProvider {
+  private readonly logger: Logger;
+  private readonly s3: S3;
+  private readonly scheduleFn: () => Promise<void>;
+  private connection?: EntityProviderConnection;
+
+  static create(options: {
+    config: AwsS3Config;
+    integration: AwsS3Integration;
+    logger: Logger;
+    schedule: TaskRunner;
+  }): AwsS3EntityProvider {
+    return new AwsS3EntityProvider(
+      options.config,
+      options.integration,
+      options.logger,
+      options.schedule,
+    );
+  }
+
+  private constructor(
+    private readonly config: AwsS3Config,
+    private readonly integration: AwsS3Integration,
+    logger: Logger,
+    schedule: TaskRunner,
+  ) {
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+
+    this.s3 = new S3({
+      apiVersion: '2006-03-01',
+      credentials: AwsCredentials.create(
+        integration.config,
+        'backstage-aws-s3-provider',
```

Comment:
Could we have `credentials: AwsCredentials` as an optional `create` and `constructor` argument, then default to this create method?

Would allow more flexibility

Reply:
Can you sketch a use case where the use of integration would not be sufficient enough?

Reply:
PS: originally I thought about adding the credential config to this plugin itself. But due to this need to be aligned with the integration due to the use of the AwsS3UrlReader down the road, I got rid of the duplication

Reply:
1. My company has a bunch of auto setup for things like role names when we deploy components, and they differ per environment. Having a single role name wouldn't work when we have multiple environments deployed to the same aws account.

2. Local dev: when working locally, we setup our aws creds outside of backstage (eg powershell environment), so there's no risk of someone accidentally committing `app-config.yaml` with creds. We roll our AwsCreds to work for both deployed and local dev options.

Reply:
All of what you describe should work I believe. We also use RBAC or temporary credentials (locally via profiles).

I assume it is about 
```
'backstage-aws-s3-provider'
```
This is only the `roleSessionName`, not a role or anything.

Reply:
Ahh ok, in which case hopefully should be fine :-)